contribute/design/1.x/cstor-operator/doc.md

```diff
@@ -544,19 +542,164 @@ spec:
+
+1. Not more than one block device should be replaced simultaneously in the same
+   raid group.
+
+2. Replacing block devices should not be already in use by the same cStor pools.
+
+3. Replacing another block device in the raid group is not allowed when any
+   of the block device in the same raid group is undergoing replacement.
+
+4. Steps to validate whether someone(other CSPC or local PV) already claimed
+   replacing block device
+
+   4.1 Verify is there any claim create for replacing block device. If there are
+       no claims for replacing block device jump to step 5.
+
+   4.2 If replacing block device was already claimed and if that claim doesn't
+       have CSPC label (or) different CSPC label(i.e block device belongs to
+       some other CSPC) then reject replacement request.
+
+   4.3 If replacing block device has self CSPC label then check is that replacing
+       block device is already linked with some other block device in that same
+       CSPC(i.e a case where users try to trigger a replace with disk that is
+       undergoing resilvering at cStor pool side). If link exists reject the request
+       else jump to step 5.
+
+5. Create a claim for replacing block device with annotation
+   `openebs.io/bd-predecessor: old_block_device_name` only if claim doesn't
+   exists.
+
+6. If claim already exists for replacing block device then update annotation
+   with proper block device name(Might be a case after creating claim update to
+   etcd got failed).
+
+`NOTE:`
+1. Pool expansion and disk replacement can go in parallel.
+
+2. Across the pool parallel block device replacements are allowed only when the
+   block devices belong to the different raid groups.
+
+Block Device Replacement Workflow: CSPC-Operator
+
+Work done by CStor-Operator after the user has updated the existing block device
+name with new block device name
+
+1. The CSPC-Operator will detect under which CSPC pool spec block device has been
+   replaced[How it can detect? By comparing new CSPC spec changes with corresponding
+   node CSPI spec] after identifying the changes CSPC-Operator will update the
+   corresponding raid group of CSPI with replacing block device changes.
+
+Work done by CSPI-Mgmt after CSPC-Operator replaces block device name in CSPI spec
+
+1. The CSPI-Mgmt will reconcile for the changes made to CSPI. CSPI-Mgmt will
+   process changes in the following manner CSPI-Mgmt will get dev links that are
+   in use by cstor pool and compare against all the devlinks of block device. If
+   dev links don't exist in cstor pool then changes are triggered for replacement
+   operation. Now CSPI-Mgmt will execute `zpool replace <pool_name> <old_disk>
+   <new_disk>` the command which will trigger disk replacement operation by cstor.
+   For every reconciliation CSPI-Mgmt will check the status of resilvering upon
+   completion of resilvering CSPI-Mgmt will remove the annotation on claim of
+   new block device and unclaim the old block device.
```

Comment:
how is the pool deletion handled? which module performs unclaim during pool deletion?

Reply:
its better if we can make sure that this unclaim will not collide with the one happens during pool deletion

Reply:
If pool is deleted then there won't be any CSPI-Mgmt guy to perform clean up process.

Reply:
Are you saying the pool label on the replaced block device?